src/compute/src/render/continual_task.rs

```diff
@@ -107,18 +111,119 @@ pub(crate) struct ContinualTaskCtx<G: Scope<Timestamp = Timestamp>> {
+                let oks = oks.flat_map(|(row, ts, diff)| {
+                    let mut negation = diff.clone();
+                    differential_dataflow::Diff::negate(&mut negation);
+                    [(row.clone(), ts.step_forward(), negation), (row, ts, diff)]
+                });
+                (oks.as_collection(), errs, times.as_collection())
+            }
+            NormalReference => {
+                let times = Collection::empty(&oks.scope());
+                (oks, errs, times)
+            }
+            // Self-references must be handled differently. When computing the
+            // proposed write to some output at `T`, we can only know the
+            // contents of it through `T-1` (the exclusive upper is `T`).
+            //
+            // We address this by initially assuming that the output contains no
+            // changes at `T`, then evaluating each of the statements in order,
+            // allowing them to see the proposed output changes made by the
+            // previous statements. By default, this is stopped after one
+            // iteration and proposed output diffs are committed if possible.
+            // (We could also add options for iterating to a fixpoint,
+            // stop/error after N iters, etc.) Then to compute the changes at
+            // `T+1`, we read in what was actually written to the output at `T`
+            // (maybe some other replica wrote something different) and begin
+            // again.
+            //
+            // The above is very similar to how timely/differential dataflow
+            // iteration works, except that our feedback loop goes through
+            // persist and the loop timestamp is already `mz_repr::Timestamp`.
+            //
+            // This is implemented as follows:
+            // - `let I = persist_source(self-reference)`
+            // - Transform `I` such that the contents at `T-1` are presented at
+            //   `T` (i.e. initially assume `T` is unchanged from `T-1`).
+            // - TODO(ct3): Actually implement the following.
+            // - In an iteration sub-scope:
+            //   - Bring `I` into the sub-scope and `let proposed = Variable`.
+            //   - We need a collection that at `(T, 0)` is always the contents
+            //     of `I` at `T`, but at `(T, 1...)` contains the proposed diffs
+            //     by the CT logic. We can construct it by concatenating `I`
+            //     with `proposed` except that we also need to retract
+            //     everything in `proposed` for the next `(T+1, 0)` (because `I`
+            //     is the source of truth for what actually committed).
+            //  - `let R = retract_at_next_outer_ts(proposed)`
+            //  - `let result = logic(concat(I, proposed, R))`
+            //  - `proposed.set(result)`
+            // - Then we return `proposed.leave()` for attempted write to
+            //    persist.
```

Comment:
Will all of this be implemented in `ContinualTaskSourceTransformer::transform`? I assume not since you don't have access to the CT statements and other inputs here. If we need to implement most of this in different places, should we make this a module-level comment instead? Makes the explanation easier to find and avoids people (like me) initially thinking that the code below does more than the `T-1` -> `T` conversion.